blockchain/contracts/HelloWorld.sol

```diff
@@ -1,52 +1,124 @@
         NFTOwnership memory nftOwnership;
         nftOwnership.nftId = nftHash;
         nftOwnership.owner = nftOwner;
-        nftOwnership.droptime = _droptime;
+        nftOwnership.reservedFor = nftOwner;
+        nftOwnership.dropTime = _dropTime;
+        nftOwnership.dropHash = _dropHash;
         nftOwnerships[nftHash] = nftOwnership;
-        nfts.push(nftHash);
-        return nftHash;
+        availableNFTs.push(nftHash);
+        nftsLength = availableNFTs.length;
+    }
+
+    function joinDrop() public {
+        // TODO: purchase limit of 5% of all available NFTs
+        require(
+            !joinedUsers.exists(msg.sender),
+            "The address you are calling from has already joined the drop."
+        );
+        // We have to make sure that not more users join the drop than we have NFTs available
+        require(joinedUsers.size() <= 10, "You cannot join the drop anymore.");
+        joinedUsers.pushAddress(msg.sender);
     }
 
-    function buyNFT(uint256 _price, uint256 _nftHash) public payable  {
-        require(nftOwnerships[_nftHash].droptime<=block.timestamp,"Droptime not yet reached!");
-        //string(abi.encodePacked("Drop has not started yet! ",((nftOwnerships[_nftHash].droptime-block.timestamp)/86400)," Days, ",((nftOwnerships[_nftHash].droptime-block.timestamp)/3600)," Hours,", ((nftOwnerships[_nftHash].droptime-block.timestamp)/60)," Minutes, and ", ((nftOwnerships[_nftHash].droptime-block.timestamp))," Seconds left.")));
+    function drop(uint256 _dropHash) public {
+        require(
+            nftOwnerships[_dropHash].dropTime <= block.timestamp,
+            "Droptime not yet reached!"
+        );
+        uint256 i;
+        uint256 nftElement;
+        uint256 nftHash;
+        for (i = 0; i < joinedUsers.size(); i++) {
+            nftElement = generateRandomNumber(i) % availableNFTs.length;
+            nftHash = availableNFTs[nftElement];
+            remove(nftElement);
+            nftOwnerships[nftHash].reservedFor = joinedUsers.getAddressAtIndex(
+                i
+            );
+        }
+    }
+
+    function buyNFT(uint256 _price, uint256 _nftHash) public payable {
+        require(
+            nftOwnerships[_nftHash].dropTime <= block.timestamp,
+            "Droptime not yet reached!"
+        );
+        require(
+            nftOwnerships[_nftHash].reservedFor == msg.sender,
+            "This NFT wasn't assigned to you"
+        );
+        // string(abi.encodePacked("Drop has not started yet! ",((nftOwnerships[_nftHash].droptime-block.timestamp)/86400)," Days, ",((nftOwnerships[_nftHash].droptime-block.timestamp)/3600)," Hours,", ((nftOwnerships[_nftHash].droptime-block.timestamp)/60)," Minutes, and ", ((nftOwnerships[_nftHash].droptime-block.timestamp))," Seconds left.")));
         _price = msg.value;
         nftOwner.transfer(_price);
-        nftOwnerships[_nftHash].owner = buyer;        
+        nftOwnerships[_nftHash].owner = buyer;
```

Comment:
Error occured while Testing: 
If I try to buy the NFT for which I was "reservedFor", transaction works, but the new "owner" is set to 0x0...0.

Reply:
This is a remix specific issue, can't do anything about it :/

Reply:
![image](https://user-images.githubusercontent.com/37510611/142289831-d8a33b00-2f43-46ac-a3a5-f084ba34fa0b.png)
found a solution :)